jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/DownloadObjectAction.kt

```diff
@@ -34,39 +64,143 @@ class DownloadObjectAction(private val project: Project, treeTable: S3TreeTable)
+            FileChooserDescriptorFactory.createSingleFileOrFolderDescriptor().also {
+                it.description = message("s3.download.object.browse.description.single")
+            }
+        }
+
+        descriptor.title = message("s3.download.object.browse.title")
+
+        // In order to prevent a confusing UX around when we can have folders and not, force the non native chooser
+        descriptor.isForcedToUseIdeaFileChooser = true
+
+        return FileChooser.chooseFile(descriptor, project, baseDir)?.path?.let {
+            Paths.get(it)
+        }
+    }
+
+    private fun checkForConflicts(targetDirectory: Path, downloads: List<DownloadInfo>): List<DownloadInfo> {
+        val finalDownloads = mutableListOf<DownloadInfo>()
+
+        for ((index, download) in downloads.withIndex()) {
+            if (!download.diskLocation.exists()) {
+                finalDownloads.add(download)
+                continue
+            }
+
+            val resolution = promptForConflictResolution(targetDirectory, download, downloads)
+            if (resolution == SKIP) {
+                LOG.info { "User requested skipping $download" }
+            } else if (resolution == OVERWRITE) {
+                finalDownloads.add(download)
+            } else if (resolution == SKIP_ALL) {
+                LOG.info { "User requested skipping rest of the files" }
+                break
+            } else if (resolution == OVERWRITE_ALL) {
+                finalDownloads.addAll(downloads.drop(index))
+                break
+            }
+        }
+
+        return finalDownloads
+    }
+
+    private fun promptForConflictResolution(
+        targetDirectory: Path,
+        download: DownloadInfo,
+        files: List<DownloadInfo>
+    ): ConflictResolution {
+        val description = message(
+            "s3.download.object.conflict.description",
+            targetDirectory.relativize(download.diskLocation),
+            targetDirectory
+        )
+
+        val options = if (files.size == 1) {
+            ConflictResolution.SINGLE_FILE_RESOLUTIONS
+        } else {
+            ConflictResolution.MULTIPLE_FILE_RESOLUTIONS
+        }
+
+        val choiceNum = Messages.showDialog(
+            project,
+            description,
+            message("s3.download.object.action"),
+            options.map { it.message }.toTypedArray(),
+            0,
+            Messages.getQuestionIcon()
+        )
+
+        return if (choiceNum < 0) {
+            SKIP
+        } else {
+            options[choiceNum]
+        }
     }
 
-    private fun downloadAll(project: Project, files: List<Pair<String, OutputStream>>) {
+    private fun downloadAll(project: Project, files: List<DownloadInfo>) {
         GlobalScope.launch {
             try {
                 files.forEach { (key, output) ->
                     try {
-                        bucket.download(project, key, output)
+                        bucket.download(project, key, output.outputStream())
```

Comment:
Can you add a TODO fix the indicator? since it makes an indicator for each file which should be fixed

Reply:
Oh ya sure